go-runtime/goplugin/service.go

```diff
@@ -173,181 +161,47 @@ func (s *Service) Build(ctx context.Context, req *connect.Request[langpb.BuildRe
-	watchTopic.Subscribe(watchEvents)
-
-	// We need watcher to calculate file hashes before we do initial build so we can detect changes
-	select {
-	case e := <-watchEvents:
-		_, ok := e.(watch.WatchEventModuleAdded)
-		if !ok {
-			return errors.Errorf("expected module added event, got: %T", e)
-		}
-	case <-time.After(3 * time.Second):
-		return errors.Errorf("expected module added event, got no event")
-	case <-ctx.Done():
-		return errors.Wrap(ctx.Err(), "context done")
-	}
+	return buildAndSend(ctx, projectConfig, req.Msg.StubsRoot, buildCtx, ongoingState, req.Msg.DevModeBuild)
 
-	go func() {
-		for e := range channels.IterContext(ctx, watchEvents) {
-			if change, ok := e.(watch.WatchEventModuleChanged); ok {
-				logger.Debugf("Found file changes: %s", change)
-				events <- updateEvent(filesUpdatedEvent{changes: change.Changes})
-			}
-		}
-	}()
-	return nil
-}
-
-// buildContextFromPendingEvents processes all pending events to determine the latest context and whether the build is automatic.
-func buildContextFromPendingEvents(ctx context.Context, buildCtx buildContext, events chan updateEvent, firstEvent updateEvent, ongoingState *compile.OngoingState) (newBuildCtx buildContext, isAutomaticRebuild bool) {
-	allEvents := []updateEvent{firstEvent}
-	// find any other events in the queue
-	for {
-		select {
-		case e := <-events:
-			allEvents = append(allEvents, e)
-		case <-ctx.Done():
-			return buildCtx, false
-		default:
-			// No more events waiting to be processed
-			hasExplicitBuilt := false
-			for _, e := range allEvents {
-				switch e := e.(type) {
-				case buildContextUpdatedEvent:
-					buildCtx = e.buildCtx
-					hasExplicitBuilt = true
-				case filesUpdatedEvent:
-					ongoingState.DetectedFileChanges(buildCtx.Config, e.changes)
-				}
-			}
-			switch e := firstEvent.(type) {
-			case buildContextUpdatedEvent:
-				buildCtx = e.buildCtx
-				hasExplicitBuilt = true
-			case filesUpdatedEvent:
-				ongoingState.DetectedFileChanges(buildCtx.Config, e.changes)
-			}
-			return buildCtx, !hasExplicitBuilt
-		}
-	}
 }
 
 // buildAndSend builds the module and sends the build event to the stream.
 //
 // Build errors are sent over the stream as a BuildFailure event.
 // This function only returns an error if events could not be send over the stream.
-func buildAndSend(ctx context.Context, stream *connect.ServerStream[langpb.BuildResponse], projectConfig projectconfig.Config, stubsRoot string, buildCtx buildContext,
-	isAutomaticRebuild bool, transaction watch.ModifyFilesTransaction, ongoingState *compile.OngoingState, devMode bool) error {
-	buildEvent, err := build(ctx, projectConfig, stubsRoot, buildCtx, isAutomaticRebuild, transaction, ongoingState, devMode)
+func buildAndSend(ctx context.Context, projectConfig projectconfig.Config, stubsRoot string, buildCtx buildContext,
+	ongoingState *compile.OngoingState, devMode bool) (*connect.Response[langpb.BuildResponse], error) {
+	captureTx := &captureTransaction{}
```

Comment:
I think the transaction stuff has to be communicated back to the engine somehow, maybe as part of the build result